Source/Peripheral.swift

```diff
@@ -245,23 +252,38 @@ public class Peripheral {
     /// - `withResponse` -  Observable emits `Next` with `Characteristic` instance write was confirmed without any errors.
     /// Immediately after that `Complete` is called. If any problem has happened, errors are emitted.
     /// - `withoutResponse` - Observable emits `Next` with `Characteristic` instance once write was called.
-    /// Immediately after that `.Complete` is called. Result of this call is not checked, so as a user you are not sure if everything completed successfully. Errors are not emitted
+    /// Immediately after that `.Complete` is called. Result of this call is not checked, so as a user you are not sure if everything completed successfully. Errors are not emitted. It ensures that peripheral is ready to write without
+    /// response by listening to the proper delegate method
     public func writeValue(_ data: Data,
                            for characteristic: Characteristic,
                            type: CBCharacteristicWriteType) -> Single<Characteristic> {
-        let observable: Observable<Characteristic>
+        let writeOperationPerformingAndListeningObservable = { [weak self] (observable: Observable<Characteristic>)
+            -> Observable<Characteristic> in
+            guard let strongSelf = self else { return Observable.error(BluetoothError.destroyed) }
+            return strongSelf.ensureValidPeripheralStateAndCallIfSucceeded(
+                for: observable,
+                postSubscriptionCall: { [weak self] in
+                    self?.peripheral.writeValue(data, for: characteristic.characteristic, type: type)
+                }
+            )
+        }
         switch type {
         case .withoutResponse:
-            observable = .just(characteristic)
+            return Observable<Characteristic>.deferred { [weak self] in
+                guard let strongSelf = self else { throw BluetoothError.destroyed }
+                let canWrite = strongSelf.monitorWriteWithoutResponseReadiness()
+                    .map { _ in true }
+                    .startWith(strongSelf.canSendWriteWithoutResponse)
+                    .filter { $0 }
+                    .take(1)
+                let observable = writeOperationPerformingAndListeningObservable(Observable.just(characteristic))
+                return canWrite
+                    .flatMap { _ in observable } 
```

Comment:
I would write above without introducing tmp variables like: canWrite and observable.

Reply:
True, I was fighting the type system. Will change it.